File: static/detector/facedet.js
/* Stage 1.5 - Haar face detection, the localization anchor for WORN glasses.
 * Port of pipeline/facedet.py.
 *
 * Pure classical CV can't find glasses on a face (hair/clothing/background
 * dominate), so we anchor on the *face* via OpenCV.js Haar cascades (Viola-Jones,
 * no neural net). Frontal (default + alt2), then a profile fallback. `locate`
 * derives the two top-outer corner ROIs from this box; no face -> held path.
 *
 * Cascades are loaded once into the OpenCV.js virtual FS by load(); the caller
 * supplies the XML bytes (fetch() in the browser, readFileSync in Node), keeping
 * this module free of any I/O specifics. Works under <script> (window.DET.facedet)
 * and Node (module.exports).
 */
(function (root, factory) {
  var mod = factory();
  if (typeof module !== 'undefined' && module.exports) { module.exports = mod; }
  if (root) { root.DET = root.DET || {}; root.DET.facedet = mod; }
})(typeof self !== 'undefined' ? self : (typeof window !== 'undefined' ? window : null),
function () {
  'use strict';

  /* Operating point mirrors facedet.py: upright only (mn=6, no rotation, +profile
     fallback) — tuned to give ZERO false faces on normal_frame while still
     catching worn faces. Adding rotated passes hallucinated faces on studio shots. */
  var SCALE_FACTOR = 1.05;
  var MIN_NEIGHBORS = 6;
  var MIN_SIZE_FRAC = 0.10;   // min face side as a fraction of min(H, W)

  var _cv = null;
  var _frontals = [];         // [CascadeClassifier] default, then alt2
  var _profile = null;        // CascadeClassifier
  var _ready = false;

  /** True once the cascades are loaded and detectFace() can run. */
  function ready() { return _ready; }

  /**
   * Load the three cascades into OpenCV.js. Call once, after cv is initialised.
   * @param {object} cv OpenCV.js module
   * @param {{frontalDefault:Uint8Array, frontalAlt2:Uint8Array, profile:Uint8Array}} bytes
   */
  function load(cv, bytes) {
    _cv = cv;
    function mk(name, data) {
      try { cv.FS_unlink(name); } catch (e) { /* not present yet */ }
      cv.FS_createDataFile('/', name, data, true, false, false);
      var c = new cv.CascadeClassifier();
      c.load(name);
      return c;
    }
    _frontals = [
      mk('fd_default.xml', bytes.frontalDefault),
      mk('fd_alt2.xml', bytes.frontalAlt2)
    ];
    _profile = mk('profileface.xml', bytes.profile);
    _ready = true;
  }

  /**
   * Largest plausible face box, or null. gray = canonical raw grayscale cv.Mat.
   * @param {cv.Mat} gray CV_8UC1
   * @param {object} [cv] OpenCV.js module (defaults to the one passed to load())
   * @returns {{x,y,w,h,method}|null}
   */
  function detectFace(gray, cv) {
    cv = cv || _cv;
    if (!_ready || !cv) { return null; }
    var H = gray.rows, W = gray.cols;
    var minSide = Math.floor(MIN_SIZE_FRAC * Math.min(H, W));
    var ms = new cv.Size(minSide, minSide);
    var maxs = new cv.Size();
    var best = null;   // {area, box:[x,y,w,h], method}

    function consider(box, method) {
      var area = box[2] * box[3];
      if (best === null || area > best.area) { best = { area: area, box: box, method: method }; }
    }

    function runOn(mat, cascade, tag, flipBack) {
      var rv = new cv.RectVector();
      cascade.detectMultiScale(mat, rv, SCALE_FACTOR, MIN_NEIGHBORS, 0, ms, maxs);
      var i;
      for (i = 0; i < rv.size(); i += 1) {
        var r = rv.get(i);
        var x = r.x;
        if (flipBack) { x = W - r.x - r.width; }
        consider([x, r.y, r.width, r.height], tag);
      }
      rv.delete();
    }

    // frontal cascades, upright
    var names = ['default', 'alt2'];
    var k;
    for (k = 0; k < _frontals.length; k += 1) {
      runOn(gray, _frontals[k], names[k] + '@0', false);
    }

    // profile fallback (image + mirror) only if nothing found yet
    if (best === null) {
      runOn(gray, _profile, 'profileL', false);
      var flipped = new cv.Mat();
      cv.flip(gray, flipped, 1);
      runOn(flipped, _profile, 'profileR', true);
      flipped.delete();
    }

    if (best === null) { return null; }
    var b = best.box;
    return { x: b[0], y: b[1], w: b[2], h: b[3], method: best.method };
  }

  return { load: load, ready: ready, detectFace: detectFace };
});

File: pipeline/facedet.py
"""Stage 1.5 - Haar face detection, the localization anchor for WORN glasses.

Pure classical CV can't find glasses on a face (hair/clothing/background always
dominate the frame), so we anchor on the *face* instead: an OpenCV Haar cascade
(Viola-Jones, no neural net, ships in OpenCV + OpenCV.js). Frontal + profile
cascades, retried over a few in-plane rotations, detect the face in every
`ray_ban_face` sample. `locate` then derives the two top-outer corner ROIs from
this box; when no face is found we fall back to the held-glasses lens path.
"""
from __future__ import annotations

import os
from dataclasses import dataclass

import cv2
import numpy as np

# Bundled cascades (single source shared with the browser port). Fall back to
# OpenCV's install dir if the bundled copies are missing.
_HAAR_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                         "models", "haar")


@dataclass
class Face:
    x: int
    y: int
    w: int
    h: int
    method: str          # which cascade/rotation found it


_CASCADE_CACHE: dict = {}


def _cascade(name: str):
    if name not in _CASCADE_CACHE:
        path = os.path.join(_HAAR_DIR, name)
        if not os.path.exists(path):
            path = cv2.data.haarcascades + name
        _CASCADE_CACHE[name] = cv2.CascadeClassifier(path)
    return _CASCADE_CACHE[name]


# Upright only: the live-webcam use case holds the glasses/face roughly level, and
# adding rotated passes hallucinated faces on studio product shots (false META on
# normal glasses). This operating point (mn=6, no rotation, +profile) was tuned to
# give ZERO false faces on normal_frame while still catching 17/20 worn faces.
ROTATIONS = (0,)
_FRONTALS = ("haarcascade_frontalface_default.xml",
             "haarcascade_frontalface_alt2.xml")
_PROFILE = "haarcascade_profileface.xml"
_SCALE_FACTOR = 1.05
_MIN_NEIGHBORS = 6
_MIN_SIZE_FRAC = 0.10        # min face side as a fraction of min(H, W)


def _rot(gray, deg):
    h, w = gray.shape[:2]
    m = cv2.getRotationMatrix2D((w / 2.0, h / 2.0), deg, 1.0)
    r = cv2.warpAffine(gray, m, (w, h), flags=cv2.INTER_LINEAR, borderValue=127)
    minv = cv2.getRotationMatrix2D((w / 2.0, h / 2.0), -deg, 1.0)
    return r, minv


def _map_box_back(box, minv):
    x, y, w, h = box
    pts = np.array([[x, y], [x + w, y], [x, y + h], [x + w, y + h]],
                   np.float32).reshape(-1, 1, 2)
    p = cv2.transform(pts, minv).reshape(-1, 2)
    x0, y0 = float(p[:, 0].min()), float(p[:, 1].min())
    x1, y1 = float(p[:, 0].max()), float(p[:, 1].max())
    return (int(round(x0)), int(round(y0)),
            int(round(x1 - x0)), int(round(y1 - y0)))


def detect_face(gray: np.ndarray) -> Face | None:
    """Largest plausible face box, or None. gray = canonical raw grayscale."""
    H, W = gray.shape[:2]
    min_side = int(_MIN_SIZE_FRAC * min(H, W))
    ms = (min_side, min_side)

    best = None                      # (area, box, method)

    def consider(box, method):
        nonlocal best
        area = box[2] * box[3]
        if best is None or area > best[0]:
            best = (area, box, method)

    # frontal cascades, upright first then rotated
    for deg in ROTATIONS:
        g, minv = _rot(gray, deg) if deg else (gray, None)
        for name in _FRONTALS:
            found = _cascade(name).detectMultiScale(g, _SCALE_FACTOR,
                                                    _MIN_NEIGHBORS, minSize=ms)
            for b in found:
                bb = _map_box_back(b, minv) if deg else tuple(int(v) for v in b)
                consider(bb, f"{name.split('_')[1]}@{deg}")
        if best is not None and deg == 0:
            break                    # good upright frontal -> stop early

    # profile fallback (image + mirror) only if nothing found yet
    if best is None:
        prof = _cascade(_PROFILE)
        for g, tag in ((gray, "profileL"), (cv2.flip(gray, 1), "profileR")):
            for b in prof.detectMultiScale(g, _SCALE_FACTOR, _MIN_NEIGHBORS,
                                            minSize=ms):
                x, y, w, h = (int(v) for v in b)
                if tag == "profileR":
                    x = W - x - w
                consider((x, y, w, h), tag)

    if best is None:
        return None
    _, (x, y, w, h), method = best
    return Face(x=x, y=y, w=w, h=h, method=method)
